### DISCUSSION TITLE
If you need interactivity, consider converting part of this to a Client Component. ????

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#67858] [Help] [answered]
### Summary

What is a Client Component ??

I am creating a new project from an old code from another project.

I created a new page in the new project like this:

app>file>page.js

Why would I be getting this error on the new project?

```
Error: Event handlers cannot be passed to Client Component props.
  <button className=... onClick={function handleSubmit} children=...>
```

^^^ I know I still need to fix the api route in the code below, but why would I not be able to include a function code???

The code for this page looks like this:

```
import React from "react";
import "bootstrap/dist/css/bootstrap.min.css";
import test from "../../thejsons/test.json";

export default function ContactForm() {
  const handleSubmit = async () => {
    const d = new Date();
    const month = d.getMonth();
    const days = d.getDate();
    const year = d.getFullYear();
    const hour = d.getHours();
    const minutes = d.getMinutes();
    const seconds = d.getSeconds();

    var timesRun = 0;
    var arrayIndex = 0;

    console.log(
      `We are moving on ${month}/${days}/${year} at ${hour}:${minutes}:${seconds}s`
    );

    var myIntervals = setInterval(async function () {
      const currentQuote = test[arrayIndex++];
      const fName = currentQuote.Contact.split(" ")[0].toLowerCase();
      const upper = fName.charAt(0).toUpperCase() + fName.slice(1);

      const data = {
        contact: currentQuote.Contact,
        email: currentQuote.Email,
        quote: currentQuote.quote,
        id: currentQuote.id,
        upper: upper,
      };

      console.log(data);

      const response = await fetch("/api/quote-email", {
        method: "POST",
        headers: {
          Accept: "application/json, text/plain, */*",
          "Content-Type": "application/json",
        },
        body: JSON.stringify(data),
      });

      let res2;

      try {
        res2 = await response.json();
      } catch (err) {
        console.log("Error parsing", err);
      }

      const resMessage = res2.message;
      console.log(resMessage);

      const d = new Date();
      const month = d.getMonth();
      const days = d.getDate();
      const year = d.getFullYear();
      const hour = d.getHours();
      const minutes = d.getMinutes();
      const seconds = d.getSeconds();
      timesRun += 1;
      console.log(
        `we ran ${timesRun} times at ${month}/${days}/${year} at ${hour}:${minutes}:${seconds}s`
      );
      console.log(arrayIndex, quotes.length);

      if (arrayIndex === quotes.length) {
        console.log("We are done!");

        clearInterval(myIntervals);
        arrayIndex = 0;
      }
    }, 10 * 1000);
  };

  return (
    <div>
      <div className="col-md-12 main-contact-form-div">
        <h4 className="get-in-touch">Get in touch.</h4>
        <br />
        <br />
        <button className="btn btn-primary form-button" onClick={handleSubmit}>
          click me
        </button>
      </div>
    </div>
  );
}




```

**Comment by icyJoseph** [comment, marked as answer]
Client Components, are the components we used up until now. Server Components are the new kid on the block.

That being said, for a matter of convenience over various subjects, Server Components, are the default. All Components are Server Components, unless:

- the file where they are defined has a `use client` directive at the top
- they are owned by a file that used `use client` directive at the top

In your example, you can just put:

```jsx
'use client'; // add this part!

import React from "react";
import "bootstrap/dist/css/bootstrap.min.css";
import test from "../../thejsons/test.json";

export default function ContactForm() {
```

This means that this component will be bundled into the JavaScript that is sent to the client! That means that it'd be able to attach event listeners, and access the window component, within certain guards.

But it also, has a complementary implication. That is that, server components don't send their JavaScript to the client, which in turn means you can ship less code!

It is important to know that Client Components, are still rendered on the server, so unguarded access to window and other DOM APIs is not recommended. Just like it was up till now.

Once again the docs are your best friend here:

- https://nextjs.org/docs/app/building-your-application/rendering
- https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#when-to-use-server-and-client-components

**Reply by icyJoseph** [reply]
@Elindo586 did you manage with this part?
